### PR TITLE
Update `getCenter()` calculation and move it to PolyUtil / LineUtil

### LINKED ISSUE #5216: Tooltip binded to polygon displacement on zoom in
## How to reproduce

Leaflet version I'm using: 1.0.2
Browser (with version) I'm using: Chrome & FF
OS/Platform (with version) I'm using: Windows 10

## What behaviour I'm expecting and which behaviour I'm seeing

Binded tooltip to polygon changes its place on zoom in. I'm attaching jsfiddle examples. 

In first example problem occurs once map is zoomed to polygon.

In second example problem does not occur if we initialize map with higher zoom.

## Minimal example reproducing the issue

https://jsfiddle.net/cco2v6vq/

https://jsfiddle.net/cco2v6vq/1/

### PR DESCRIPTION
**Pull request by Falke-Design**
Changed the `getCenter()` function for Polygon & Polyline. Now it uses `map.un-/project()` of the latlngs instead of the pre-generated `_rings` with `latLngToLayerPoint()`.

Also added the center functions to `PolyUtil` and `LineUtil`, because then the function can be used without creating a Polyline/-gon and adding it to the map:
grafik

## Visible changes:

**Polyline**:
 - if `getCenter()` is called on a low zoom level (0), then the returned point is not anymore the first latlng:
grafik
- `getCenter()` returns the same point no matter which zoom level (2 & 16)
grafik

**Polygon**:
 - if `getCenter()` is called on a low zoom level (0), then the returned point is not anymore the first latlng:
grafik
- `getCenter()` returns the same point no matter which zoom level (8 & 16)
grafik

Fix #5216

### CODE REVIEW

**Comment by johnd0e** [comment]
Your solution is correct.
The only unclear thing is performance difference.

**Comment by Falke-Design** [comment]
I have done some tests, it looks like the new calculation is faster:
Old one:
```
Count: 100.000
performance: 4440ms //Rectangle
performance: 8931ms //Polygon
performance: 12969ms //Polyline

performance: 4443ms
performance: 8935ms
performance: 13094ms

performance: 4136ms
performance: 8532ms
performance: 12389ms

Count: 1.000.000
performance: 47880ms
performance: 97666ms
performance: 145388ms

performance: 47784ms
performance: 98649ms
performance: 144880ms
```
New:
```
Count: 100.000
performance: 4634ms
performance: 8484ms
performance: 12012ms

performance: 3850ms
performance: 7774ms
performance: 11199ms

performance: 4365ms
performance: 8241ms
performance: 11629ms

Count: 1.000.000
performance: 39544ms
performance: 84632ms
performance: 126625ms

performance: 39555ms
performance: 84180ms
performance: 125914ms
```

Code:
```
window.onload=()=> {
        setTimeout(() => {

          var COUNT = 1000*1000;

          console.log('start')
          console.time('performance');
          for (var i = 0; i < COUNT; i++) {
            var rect = L.rectangle([[48.78232, 9.17702], [48.8, 9.2]]).addTo(map);
            rect.getCenter();
            rect.removeFrom(map);
          }
          console.timeLog('performance');
          for (var i = 0; i < COUNT; i++) {
            var rect = L.polygon([[48.78232, 9.17702], [48.8, 9.2], [48.7, 9.19], [48.7, 9.13]]).addTo(map);
            rect.getCenter();
            rect.removeFrom(map);
          }
          console.timeLog('performance');
          for (var i = 0; i < COUNT; i++) {
            var rect = L.polyline([[48.78232, 9.17702], [48.8, 9.2], [48.7, 9.19], [48.7, 9.13]]).addTo(map);
            rect.getCenter();
            rect.removeFrom(map);
          }
          console.timeEnd('performance')

        }, 2000)
      }
```

File: src/geometry/LineUtil.js
Comment (johnd0e) [inline src/geometry/LineUtil.js:265]: 1. Excessive call of `map.getMaxZoom()`
2. Do you have tests with `Infinity` zoom?
3. Do you have tests with `undefined` zoom?

File: src/geometry/LineUtil.js
Comment (johnd0e) [inline src/geometry/LineUtil.js:249]: Not sure that we are ready to expose that functions, without any discussion.
- As it is (relatively) low-level function, what if we pass `crs` and `zoom` directly, instead of map?
- What if we give reasonable defaults for `crs` and `zoom`?

File: src/geometry/LineUtil.js
Comment (johnd0e) [inline src/geometry/LineUtil.js:270]: Isn't that the same value as `latlngs.length`?

File: src/geometry/LineUtil.js
Comment (johnd0e) [inline src/geometry/LineUtil.js:263]: Why do we need explicit throw here?
It will throw anyway on next line.

File: src/layer/vector/Polygon.js
Comment (johnd0e) [inline src/layer/vector/Polygon.js:70]: Again, it will throw naturally anyway.
You can put that on the test, by the way.

**Review by johnd0e — COMMENTED** [review]
It seems impossible, but I can confirm that new code is faster.
I have no idea why that could be, considering that new code uses more `project` calls.

File: src/geometry/LineUtil.js
Comment (Falke-Design) [reply src/geometry/LineUtil.js:270]: I copied this from the original code. But I think it is done because `latlngs.length` returns not the correct length if it is a multipolyline.

File: src/geometry/LineUtil.js
Comment (Falke-Design) [inline src/geometry/LineUtil.js:256]: I need to check this, in the original function was allowed to pass a multiployline

File: src/geometry/LineUtil.js
Comment (Falke-Design) [reply src/geometry/LineUtil.js:249]: I think this is a good idea

File: src/geometry/PolyUtil.js
Comment (Falke-Design) [inline src/geometry/PolyUtil.js:67]: Should be possible to use for a multipolygon too

File: src/geometry/LineUtil.js
Comment (Falke-Design) [reply src/geometry/LineUtil.js:263]: Copied from original `getCenter()`. It is clearer but it is not necessary

File: src/layer/vector/Polygon.js
Comment (Falke-Design) [reply src/layer/vector/Polygon.js:70]: Clearer message, if it throw only `this._map is undefined` it is not clear why this error happens

Comment (johnd0e) [reply src/layer/vector/Polygon.js:70]: > it is not clear why this error happens

But it gets crystal clear when you see traceback.

There are A LOT of such places in leaflet, and if we age going to explicitly check every one - better switch to typescript instead.

File: src/geometry/LineUtil.js
Comment (johnd0e) [reply src/geometry/LineUtil.js:270]: > I copied this from the original code.

But it is redundant for the new code, right?

File: src/layer/vector/Polygon.js
Comment (Falke-Design) [reply src/layer/vector/Polygon.js:70]: I don't totally agree with you. This check here makes sense because it would not be clear why `getCenter()` is not working. It is a common problem that users want to use `getCenter()` before adding to the map.

File: src/geometry/LineUtil.js
Comment (Falke-Design) [reply src/geometry/LineUtil.js:270]: Yep if 
> But I think it was / is used because `latlngs.length` returns not the correct length if it is a multipolyline.

makes no problems

File: src/layer/vector/Polygon.js
Comment (johnd0e) [reply src/layer/vector/Polygon.js:70]: > It is a common problem that users want to use `getCenter()` before adding to the map.

So? It fails, you see (short) call stack, you see args, and all gets clear. No?

File: src/geometry/LineUtil.js
Comment (johnd0e) [reply src/geometry/LineUtil.js:270]: But here latlngs mapped to points directly, so I do not see any difference.

Comment (Falke-Design) [reply src/geometry/LineUtil.js:270]: I need to check this mapping because this is not working for mult-lines. Currently `latlngs.length` work because I check with `if (!isFlat(latlngs)) {` if it is a multi-line. If we say that multipolylines should supported too, it will not work.

Can it be that the old `getCenter()` is also only working for **no**t mulit-lines? Because it uses only the first array `points = this._rings[0],`

File: src/geometry/LineUtil.js
Comment (johnd0e) [reply src/geometry/LineUtil.js:256]: Instead we could document that only first ring is processed for multi-features.

File: src/geometry/LineUtil.js
Comment (Falke-Design) [reply src/geometry/LineUtil.js]: 1. replaced - reading out once and then check in the if
2. The most tests are with maxZoom Infinity, or do you mean the actual zoom `getZoom()`? When does this happen?
3. When does this happen?

**Comment by Falke-Design** [comment]
Updated
- Change arguments to `crs` and `zoom` instead of `map`.
- Allow non flat latlngs but only takeing the first ring

File: src/layer/vector/Polygon.js
Comment (Falke-Design) [reply src/layer/vector/Polygon.js:70]: @mourner what do you think? Should we remove this error message from Polygon & Polyline in the `getCenter()` method?

File: spec/suites/geometry/LineUtilSpec.js
Comment (johnd0e) [inline spec/suites/geometry/LineUtilSpec.js:90]:
```suggestion
			map = L.map(document.createElement('div'), {center: [55.8, 37.6], zoom: 6});
```
(Same for polygon)

File: spec/suites/geometry/LineUtilSpec.js
Comment (johnd0e) [inline spec/suites/geometry/LineUtilSpec.js:101]: Too long line (here and below)

File: spec/suites/geometry/LineUtilSpec.js
Comment (johnd0e) [inline spec/suites/geometry/LineUtilSpec.js:107]: What if define `zoom` and `crs` variables in `before` hook?

File: spec/suites/geometry/LineUtilSpec.js
Comment (johnd0e) [inline spec/suites/geometry/LineUtilSpec.js:141]: Should be before `expect`

File: spec/suites/geometry/PolyUtilSpec.js
Comment (johnd0e) [inline spec/suites/geometry/PolyUtilSpec.js:93]: Should be before `expect`

File: spec/suites/layer/vector/PolygonSpec.js
Comment (johnd0e) [inline spec/suites/layer/vector/PolygonSpec.js:173]: You should then restore the zoom, here or in `afterEach` hook.

File: spec/suites/layer/vector/PolygonSpec.js
Comment (johnd0e) [inline spec/suites/layer/vector/PolygonSpec.js:194]: Same here, and in all other places.

File: spec/suites/layer/vector/PolygonSpec.js
Comment (johnd0e) [inline spec/suites/layer/vector/PolygonSpec.js:204]: Why are you using different precision values?
What if you stick to default?

**Review by johnd0e — CHANGES_REQUESTED** [review]

**Comment by johnd0e** [comment]
Just for info: there are other ways to find center of polygon:
https://blog.mapbox.com/a-new-algorithm-for-finding-a-visual-center-of-a-polygon-7c77e6492fbc

File: spec/suites/layer/vector/PolygonSpec.js
Comment (Falke-Design) [reply spec/suites/layer/vector/PolygonSpec.js]: The latlng values are long numbers but I replaced it now with the exact number, it is cleaner

File: spec/suites/layer/vector/PolygonSpec.js
Comment (Falke-Design) [reply spec/suites/layer/vector/PolygonSpec.js:194]: I changed it to `beaforeEach` so for each test the map is new created and the zoom & center will be set

File: spec/suites/layer/vector/PolygonSpec.js
Comment (Falke-Design) [reply spec/suites/layer/vector/PolygonSpec.js:173]: I changed it to `beaforeEach` so for each test the map is new created and the zoom & center will be set

**Comment by Falke-Design** [comment]
> Just for info: there are other ways to find center of polygon:
https://blog.mapbox.com/a-new-algorithm-for-finding-a-visual-center-of-a-polygon-7c77e6492fbc

Very interessting. It would be nice to have this but if we use the code from Mapbox we need to add the copyright. I would create it from the explanation but I'm not good in algorithm / mathematical things.
I suggest we merge this PR and then we can look if we want to use the "visual center algorithm".

File: spec/suites/geometry/PolyUtilSpec.js
Comment (johnd0e) [inline spec/suites/geometry/PolyUtilSpec.js:88]:
```suggestion
			expect(spy.calledOnce).to.be.ok();
```

File: spec/suites/geometry/LineUtilSpec.js
Comment (johnd0e) [inline spec/suites/geometry/LineUtilSpec.js:140]:
```suggestion
			expect(spy.calledOnce).to.be.ok();
```

File: src/layer/vector/Polyline.js
Comment (johnd0e) [inline src/layer/vector/Polyline.js:117]: Why remove that?

**Review by johnd0e — DISMISSED** [review]

File: src/layer/vector/Polyline.js
Comment (Falke-Design) [reply src/layer/vector/Polyline.js]: Reverted

**Comment by johnd0e** [comment]
In my opinion it is ready to merge, except 1 unresolved question: do we need explicit throws (see above).
@mourner @IvanSanchez @jonkoops

**Comment by johnd0e** [comment]
@Falke-Design 
What about numerous `.nearLatLng(L.latLng...` occurrences?

**Review by jonkoops — APPROVED** [review]

**Comment by jonkoops** [comment]
This change is looking good to me. @johnd0e do you feel comfortable merging in this change?

**Comment by Falke-Design** [comment]
@jonkoops I don't think that @johnd0e will response, he was offline since january. I would like to merge it now, because we have now 2 approvals with only one question "should we keep the already existing exceptions?". I think they don't hurt and else we can still remove them.

**Comment by jonkoops** [comment]
Yes I agree. Let's go ahead and merge this in 👍

**Pull request merged**
